### DISCUSSION TITLE
Do I need to implement all application code and logic in ViewModel files?

### DISCUSSION THREAD

**Discussion by al-zakaria** [AvaloniaUI/Avalonia#15889] [Q&A]
Hello.  I would like to ask a few questions.  1. Is it necessary to write all C# code and, in general, all application logic in ViewModel files?  I ask this because I do all the code in my applications in ViewModel files and always use MVVM and Community Toolkit when writing in Avalonia.  I ask this because there are also .axaml.cs files.  Do I need to implement some part of the code in .axaml cs files or am I doing everything correctly when I write all the code in ViewModel files?  Question 2 is how to understand that my Avalonia code is bad?  I understand that there are standards for writing code in C#, for example, do not duplicate code or use magic numbers, and I try not to do this and make many methods to avoid code duplication and make many fields and variables so that there are no magic numbers, but are there any  indicators in Avalonia that my code is bad or not?

sorry for my English.  This is not my native language and I probably made mistakes when writing the question.

**Comment by thevortexcloud** [comment]
> s it necessary to write all C# code and, in general, all application logic in ViewModel files? 

If you are following MVVM properly, then actually your main logic should not be in the view models at all. The `M` in MVVM stands for Model, which is where your actual application logic should be. The View Model wraps it and makes it available for the view to consume/interact with. As such putting _everything_ in the VM is actually incorrect. To quote Microsoft: 

> Model classes are non-visual classes that encapsulate the app's data. Therefore, the model can be thought of as representing the app's domain model, which usually includes a data model along with business and validation logic. Examples of model objects include data transfer objects (DTOs), Plain Old CLR Objects (POCOs), and generated entity and proxy objects.
>
> Model classes are typically used in conjunction with services or repositories that encapsulate data access and caching.

That said a lot of people don't follow MVVM properly anyway and just do everything all over the place/in the wrong place.

> Do I need to implement some part of the code in .axaml cs files

Depends on what you are doing. For some things it makes sense to it in the code behind since it's a view only concern. In some cases there is no straightforward way to get data to your View Model without using code behind as well. In MVVM, the View is allowed to interact with the view model. The View Model however is not allowed to directly interact with the view. EG If I need an event that is only accessible in the view, sometimes I will just manually make it execute a command/call a method in the view model when the event fires as there is no direct way to get the event information in the view model.

> am I doing everything correctly when I write all the code in ViewModel files?

Generally, you should put as much as your logic in your model and View Model classes as you can. This allows for greater portability of your code between other UIs and frameworks. As the view code will always be UI specific (in this case, Avalonia specific).

> but are there any indicators in Avalonia that my code is bad or not?

This is subjective and depends on what patterns you are following, and what you are trying to make. I typically try to keep things as consistent as I can and also as simple as I can. I also try to follow MVVM when I can since it makes things a lot easier with Avalonia/XAML frameworks.

But Avalonia does not actually require you to use MVVM at all. It just makes it much easier to use. There is nothing stopping you from putting everything into the code behind.

**Reply by al-zakaria** [reply]
I didn't really understand.  you say that according to the rules of the MVVM template, there should be no code at all in ViewModel files.  and then say that it depends on what I write.  or did I misunderstand?

**Reply by thevortexcloud** [reply]
> here should be no code at all in ViewModel files.

That's not exactly what I said. The view model is primarily responsible for preparing your data for presentation as well as informs the view about any changes to the data. To quote Microsoft (emphasis is mine):

> The view model implements properties and commands to which the view can data bind to, and notifies the view of any state changes through change notification events. The properties and commands that the view model provides define the functionality to be offered by the UI, but the view determines how that functionality is to be displayed.
>
> **The view model is also responsible for coordinating the view's interactions with any model classes that are required. There's typically a one-to-many relationship between the view model and the model classes.** The view model might choose to expose model classes directly to the view so that controls in the view can data bind directly to them. In this case, the model classes will need to be designed to support data binding and change notification events.
>
> **Each view model provides data from a model in a form that the view can easily consume.** To accomplish this, the view model sometimes performs data conversion. Placing this data conversion in the view model is a good idea because it provides properties that the view can bind to. For example, the view model might combine the values of two properties to make it easier to display by the view.

Your main application logic should be in your model classes. Your logic for what data to display should be in your view model. And your presentation logic should be in your view. Your view model can contain logic for interacting with your model classes. For example take the following class: 

```csharp
public class MyModel {
  public IEnumerable<string> GetData() {
  ...
  }
}
```

A view model would be allowed to call the `MyData` method to get a list of data to place into an `ObservableCollection` when a `ICommand` is executed. But the view model itself does not care how `GetData` works since that's not the view models job. The View however would never call `GetData` itself directly. Instead it would rely on the View Model to do that.

>  and then say that it depends on what I write

As I said, it's not possible to do everything in the view model without some messy hacks/breaking MVVM. As such it really depends on what you are doing. Generally speaking, your view models and your models should not even know Avalonia exists. But most people also don't strictly follow MVVM (even if they should) either out of laziness, convenience, or not wanting to deal with strict MVVM. I have seen a lot of projects that stick pretty much everything into the View Model despite that not being the correct place to put a lot of the logic.

If you follow a dependency injection pattern, you will mostly incidentally end up doing MVVM correctly anyway.

**Reply by al-zakaria** [reply]
I understand.  so is it better to split the code between ViewModel and .axaml.cs?

**Reply by thevortexcloud** [reply]
It depends? Generally you should avoid putting things in the code behind if you are following MVVM. As it reduces portability of your code, which is one of the main points of MVVM. But splitting it should not be seen as a bad thing if there is not a better way to do something.

**Reply by al-zakaria** [reply]
Sorry for saying so much, but every time I get more and more confused.  Now you are saying that there is no need to write code in the code behind.  but the code behind is the .axaml.cs files?  but then it turns out that the code needs to be written in ViewModel files.  or again I misunderstood

**Reply by thevortexcloud** [reply]
* Code behind is for code that is view specific (EG click events). View specific code also includes things that are specific to the UI framework such as Avalonia.
* View model is for the view to bind to and to get data from
* The model is where the view model gets data from

It is possible to do things in the code behind (such as subscribe to view events like `Loaded` and `Click`) that you can't do in a view model as the view model is not the view. Similarly, it makes more sense to do some things entirely in the code behind if they have no impact on the view model (EG controlling animations). In those cases you should use the code behind.

**MVVM allows for using the code behind, view model and models all at once.** I would expect code to be written in every one of those places in a lot of apps. Most simple apps however only need view models, models, and XAML views. The reason I said you should avoid using code behind is makes it very hard to reuse your code outside Avalonia (EG if you want to port your app to MAUI one day). But you can't always avoid it.

It might help if you watch/read a WPF/MVVM tutorial. Avalonia works _very_ similarly to WPF.
